### PR TITLE
[v7r2] Introduce Local SubLogger

### PR DESCRIPTION
**Pull request by chaen**
This PR addresses https://github.com/DIRACGrid/DIRAC/issues/5280

The meat of this PR is in b473d1a , and I apply it as a test to `JobStateUpdate` in b473d1a 

BEGINRELEASENOTES
*Framework
NEW: Introduce Local SubLogger

*WMS
CHANGE: attempt to use Local subLogger

ENDRELEASENOTES

### CODE REVIEW

File: src/DIRAC/FrameworkSystem/private/standardLogging/Logging.py [hunk 1]
```diff
@@ -433,6 +474,8 @@ def getSubLogger(self, subName, child=True):
     """
     Create a new Logging object, child of this Logging, if it does not exists.
 
+    :warning: for very short lived sub logger, consider :py:meth`getLocalSubLogger` instead
```
Comment (andresailer) [inline src/DIRAC/FrameworkSystem/private/standardLogging/Logging.py:477]:
```suggestion
    :warning: for very short lived sub logger, consider :py:meth:`getLocalSubLogger` instead
```

File: src/DIRAC/FrameworkSystem/private/standardLogging/Logging.py [hunk 2]
```diff
@@ -446,7 +489,50 @@ def getSubLogger(self, subName, child=True):
...
       return childLogging
     finally:
       self._lockInit.release()
+
+  def getLocalSubLogger(self, subName):
+    """
+        Create a subLogger which is meant to have very short lifetime,
+        (e.g. when you want to add the jobID in the name)
+
+        :warning: this is a light version of a logger, read the documentation of
```
Comment (andresailer) [inline src/DIRAC/FrameworkSystem/private/standardLogging/Logging.py:503]:
```suggestion
        .. warning::
          this is a light version of a logger, read the documentation of
```

File: src/DIRAC/FrameworkSystem/private/standardLogging/Logging.py [hunk 3]
```diff
@@ -446,7 +489,50 @@ def getSubLogger(self, subName, child=True):
...
+        :py:class:`CustomLoggingAdapter` carefully
```
Comment (andresailer) [inline src/DIRAC/FrameworkSystem/private/standardLogging/Logging.py:504]:
```suggestion
          :py:class:`CustomLoggingAdapter` carefully
```

**Comment by andresailer** [comment]
```
/home/docs/checkouts/readthedocs.org/user_builds/dirac/checkouts/5284/src/DIRAC/FrameworkSystem/private/standardLogging/Logging.py:docstring of DIRAC.FrameworkSystem.private.standardLogging.Logging.Logging.getLocalSubLogger:5: WARNING: Field list ends without a blank line; unexpected unindent.
/home/docs/checkouts/readthedocs.org/user_builds/dirac/checkouts/5284/src/DIRAC/FrameworkSystem/private/standardLogging/LoggingRoot.py:docstring of DIRAC.FrameworkSystem.private.standardLogging.Logging.Logging.getLocalSubLogger:5: WARNING: Field list ends without a blank line; unexpected unindent.
```

**Comment by andresailer** [comment]
Can you give an example how localContext is to be used? And mention this in the release notes.

**Comment by chaen** [comment]
> Can you give an example how localContext is to be used? And mention this in the release notes.

Actually, I don't think it should be used, except in the case of the `LocalSubLogger`. I wanted to avoid copy paste, that's why I added the `local_context` argument to all the `info`, `debug`, etc.
But maybe I should just go for it, and only have the `local_context` in the `_createLogRecord` method. Any thought ?

**Comment by andresailer** [comment]
I think, If it shouldn't be used in the public functions then it shouldn't be there as an argument.

**Comment by chaen** [comment]
> I think, If it shouldn't be used in the public functions then it shouldn't be there as an argument.

very valid argument, I'll change it

**Comment by chaen** [comment]
@andresailer better like this ?

File: src/DIRAC/FrameworkSystem/private/standardLogging/Logging.py [hunk 4]
```diff
@@ -325,63 +342,81 @@ def getAllPossibleLevels():
     """
     return LogLevels.getLevelNames()
 
-  def always(self, sMsg, sVarMsg=''):
+  def always(self, sMsg, sVarMsg='', local_context=None):
```
Comment (andresailer) [inline src/DIRAC/FrameworkSystem/private/standardLogging/Logging.py:345]: You still want to remove these again, right?

Comment (chaen) [reply src/DIRAC/FrameworkSystem/private/standardLogging/Logging.py:345]: oh ! something went wrong, I removed it for sure, I must have screwed my squash....

File: src/DIRAC/FrameworkSystem/private/standardLogging/Logging.py [hunk 5]
```diff
@@ -433,6 +457,10 @@ def getSubLogger(self, subName, child=True):
     """
     Create a new Logging object, child of this Logging, if it does not exists.
 
+    .. warning::
+
+       For very short lived sub logger, consider :py:meth`getLocalSubLogger` instead
```
Comment (andresailer) [inline src/DIRAC/FrameworkSystem/private/standardLogging/Logging.py:462]:
```suggestion
       For very short lived sub logger, consider :py:meth:`getLocalSubLogger` instead
```

File: src/DIRAC/FrameworkSystem/private/standardLogging/Formatter/BaseFormatter.py [hunk 6]
```diff
@@ -51,7 +51,16 @@ def format(self, record):
       if record.timeStampIsShown:
         timeStamp = '%(asctime)s UTC '
       if record.contextIsShown:
-        contextComponent = '%(componentname)s%(customname)s '
+        contextComponentList = ['%(componentname)s%(customname)s']
+
+        # The local_name is normally only provided when using a
+        # local sub logger relying on
+        # :py:class:`DIRAC.FrameworkSystem.private.standardLogging.Logging.CustomLoggingAdapter`
```
Comment (andresailer) [inline src/DIRAC/FrameworkSystem/private/standardLogging/Formatter/BaseFormatter.py:58]:
```suggestion
        # :py:class:`DIRAC.FrameworkSystem.private.standardLogging.Logging.LocalSubLogger`
```
Right?

File: src/DIRAC/FrameworkSystem/private/standardLogging/Logging.py [hunk 7]
```diff
@@ -446,7 +474,138 @@ def getSubLogger(self, subName, child=True):
...
+          local_context=self.extra)
+
+  def getLocalSubLogger(self, subName):
+    """
+        Create a subLogger which is meant to have very short lifetime,
+        (e.g. when you want to add the jobID in the name)
+
+        .. warning::
+          This is a light version of a logger, read the documentation of
+          :py:class:`CustomLoggingAdapter` carefully
```
Comment (andresailer) [inline src/DIRAC/FrameworkSystem/private/standardLogging/Logging.py:606]:
```suggestion
          :py:class:`LocalSubLogger` carefully
```

File: src/DIRAC/FrameworkSystem/private/standardLogging/Logging.py (hunk 5 above)
Comment (chaen) [reply src/DIRAC/FrameworkSystem/private/standardLogging/Logging.py:462]: done, thanks

File: src/DIRAC/FrameworkSystem/private/standardLogging/Formatter/BaseFormatter.py (hunk 6 above)
Comment (chaen) [reply src/DIRAC/FrameworkSystem/private/standardLogging/Formatter/BaseFormatter.py:58]: Almost, `Logging.Logging.LocalSubLogger` :-)

File: src/DIRAC/FrameworkSystem/private/standardLogging/Logging.py (hunk 7 above)
Comment (chaen) [reply src/DIRAC/FrameworkSystem/private/standardLogging/Logging.py:606]: done thanks

**Review by andresailer — APPROVED** [review]

**Review by fstagni — APPROVED** [review]

**Pull request merged**